regression after usingnamespace changes - termios broken when using pThreads

It is no longer possible to use termios

/usr/lib/zig/std/os.zig:156:27: error: container 'std.c' has no member called 'termios'
pub const termios = system.termios;

That does not solve the problem.   Building outwith pThreads I get (the
errors in Condition.zig are expected):

/home/ed/zig/zbox/src/prim.zig:242:24: error: container 'std.os' has no
member called 'ISIG'
   termios.lflag |= os.ISIG;
                      ^
/home/ed/zig/zbox/src/prim.zig:151:32: error: container 'std.os' has no
member called 'winsize'
   var winsize = mem.zeroes(os.winsize);
                              ^
/home/ed/zig/zbox/src/prim.zig:199:11: error: container 'std.os' has no
member called 'tcflag_t'
       os.tcflag_t,
         ^
/usr/lib/zig/std/Thread/Condition.zig:154:14: error: no member named
'unlock' in struct 'std.Thread.Mutex'
       mutex.unlock();
            ^
/usr/lib/zig/std/Thread/Condition.zig:130:30: error: container
'std.os.linux' has no member called 'FUTEX_PRIVATE_FLAG'
                       linux.FUTEX_PRIVATE_FLAG | linux.FUTEX_WAKE,

Looks like flags are missing in std.os.  When building with pThreads I get:

/usr/lib/zig/std/os.zig:157:27: error: container 'std.c' has no member
called 'termios'
pub const termios = system.termios;
                         ^
/home/ed/zig/zbox/src/prim.zig:309:25: note: referenced here
   original_termios: os.termios = undefined,
                       ^
/home/ed/zig/zbox/src/prim.zig:229:20: note: referenced here
   const handle = state().tty.in.context.handle;
                  ^
/home/ed/zig/zbox/src/prim.zig:151:32: error: container 'std.os' has no
member called 'winsize'
   var winsize = mem.zeroes(os.winsize);

Now std.os.zig is getting an error with std.c missing a member and again it
looks like std.os is missing fields.

My program uses zbox ( https://github.com/jessrud/zbox ) with fixes below
applied.  Its possible that more changes are needed in zbox but,
atleast to me, it looks like zig has a problem with termios when using
pthreads and std.os is missing some fields.  Please let me know if I am
mistaken.

Thanks
Ed

diff --git a/src/box.zig b/src/box.zig
index 72c380a..08ca207 100644
--- a/src/box.zig
+++ b/src/box.zig
@@ -329,6 +329,8 @@ pub const Buffer = struct {
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) @TypeOf(writer).Error!void {
+        _ = fmt;
+        _ = options;
        var row_num: usize = 0;
        try writer.print("\n\x1B[4m|", .{});

diff --git a/src/prim.zig b/src/prim.zig
index c809f78..8cef838 100644
--- a/src/prim.zig
+++ b/src/prim.zig
@@ -1,3 +1,8 @@
+//! the primitive terminal module is mainly responsible for providing a
simple
+//! and portable interface to pseudo terminal IO and control primitives to
+//! higher level modules. You probably shouldn't be using this directly
from
+//! application code.
+
const std = @import("std");
const fs = std.fs;
const os = std.os;
@@ -11,11 +16,6 @@ const Allocator = mem.Allocator;

usingnamespace @import("util.zig");

-//! the primitive terminal module is mainly responsible for providing a
simple
-//! and portable interface to pseudo terminal IO and control primitives to
-//! higher level modules. You probably shouldn't be using this directly
from
-//! application code.
-
/// Input events
pub const Event = union(enum) {
    tick,
@@ -150,9 +150,9 @@ const TermSize = struct { height: usize, width: usize
};
pub fn size() os.UnexpectedError!TermSize {
    var winsize = mem.zeroes(os.winsize);
    const err = os.system.ioctl(state().tty.in.context.handle,
os.TIOCGWINSZ, @ptrToInt(&winsize));
-    if (os.errno(err) == 0)
+    if (os.errno(err) == .SUCCESS)
        return TermSize{ .height = winsize.ws_row, .width = winsize.ws_col
};
-    return os.unexpectedErrno(err);
+    return os.unexpectedErrno(os.errno(err));
}

/// Hides cursor if visible
@@ -291,7 +291,7 @@ pub fn nextEvent() (Allocator.Error ||
ErrorSet.TtyRead)!?Event {
        }
    }
    const event = parseEvent();
-    debug("event: {}", .{event});
+    std.log.debug("event: {}", .{event});
    return event;
}










On Mon, Sep 13, 2021 at 1:41 PM Sage Hane ***@***.***> wrote:

> std.c.termios was renamed to std.os.termios as of c05a20f
> <https://github.com/ziglang/zig/commit/c05a20fc8c36742dab8792d15e79716da1a55759>.
> Hope that addresses your concern.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ziglang/zig/issues/9707#issuecomment-918423197>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAS6CFPEKMDQ7EJI2T6O23UBYZS7ANCNFSM5DVLVVNQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

I don't quite get what you're trying to do, but this part seems solvable:
```
/home/ed/zig/zbox/src/prim.zig:242:24: error: container 'std.os' has no member called 'ISIG'
   termios.lflag |= os.ISIG;
                      ^
/home/ed/zig/zbox/src/prim.zig:151:32: error: container 'std.os' has no member called 'winsize'
   var winsize = mem.zeroes(os.winsize);
                              ^
/home/ed/zig/zbox/src/prim.zig:199:11: error: container 'std.os' has no member called 'tcflag_t'
       os.tcflag_t,
         ^
```
>Looks like flags are missing in std.os (think they are moved to std.os.linux).

Yep, that seems correct.
Looking at https://github.com/ziglang/zig/blob/master/lib/std/os.zig#L55,
`std.os.system` should provide the members you're looking for.

The other errors seem unrelated to this issue?

--- a/lib/std/c/linux.zig
+++ b/lib/std/c/linux.zig
@@ -90,6 +90,7 @@ pub const sigset_t = linux.sigset_t;
 pub const sockaddr = linux.sockaddr;
 pub const socklen_t = linux.socklen_t;
 pub const stack_t = linux.stack_t;
+pub const termios = linux.termios;
 pub const time_t = linux.time_t;
 pub const timespec = linux.timespec;
 pub const timeval = linux.timeval;

